sdk/lib/atl/atlwin.h

```diff
@@ -1276,20 +1279,180 @@ class CWindow
 
 __declspec(selectany) RECT CWindow::rcDefault = { CW_USEDEFAULT, CW_USEDEFAULT, 0, 0 };
 
-template <class TBase = CWindow, class TWinTraits = CControlWinTraits>
-class CWindowImplBaseT : public TBase, public CMessageMap
+template <class TBase = CWindow>
+class CWindowImplRoot : public TBase, public CMessageMap
 {
 public:
     enum { WINSTATE_DESTROYED = 0x00000001 };
-    DWORD m_dwState;
-    const _ATL_MSG *m_pCurrentMsg;
+
+public:
     CWndProcThunk m_thunk;
+    const _ATL_MSG *m_pCurrentMsg;
+    DWORD m_dwState;
+
+    CWindowImplRoot()
+        : m_pCurrentMsg(NULL)
+        , m_dwState(0)
+    {
+    }
+
+    virtual ~CWindowImplRoot()
+    {
+    }
+};
+
+
+template <class TBase = CWindow>
+class CDialogImplBaseT : public CWindowImplRoot<TBase>
+{
+public:
+    // + Hacks for gcc
+    using CWindowImplRoot<TBase>::WINSTATE_DESTROYED;
+    // - Hacks for gcc
+
+    virtual ~CDialogImplBaseT()
+    {
+    }
+    virtual DLGPROC GetDialogProc()
+    {
+        return DialogProc;
+    }
+
+    static INT_PTR CALLBACK StartDialogProc(HWND hWnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
+    {
+        CDialogImplBaseT<TBase> *pThis;
+        DLGPROC newDlgProc;
+        DLGPROC GCCU(pOldProc);
+
+        pThis = reinterpret_cast<CDialogImplBaseT<TBase>*>(_AtlWinModule.ExtractCreateWndData());
+        ATLASSERT(pThis != NULL);
+        if (pThis == NULL)
+            return 0;
+        pThis->m_thunk.Init((WNDPROC)pThis->GetDialogProc(), pThis);
+        newDlgProc = (DLGPROC)pThis->m_thunk.GetWNDPROC();
```

Comment:
Should be `static_cast`? You use them elsewhere in this file.

Reply:
yup